File: frontend/src/services/notification-preferences.ts

import { api } from "./api";

/* ============================================================
   TYPES
============================================================ */

export interface NotificationPreferences {
  task_assigned: boolean;
  status_changed: boolean;
  task_created: boolean;
  task_updated: boolean;
  mention: boolean;
  comment_added: boolean;
}

export interface NotificationPreferencesUpdate {
  task_assigned?: boolean;
  status_changed?: boolean;
  task_created?: boolean;
  task_updated?: boolean;
  mention?: boolean;
  comment_added?: boolean;
}

/* ============================================================
   GET PREFERENCES
============================================================ */

export async function getNotificationPreferences(): Promise<NotificationPreferences> {
  const response = await api.get<NotificationPreferences>(
    "/api/v1/notification-preferences",
  );
  return response.data;
}

/* ============================================================
   UPDATE PREFERENCES
============================================================ */

export async function updateNotificationPreferences(
  data: NotificationPreferencesUpdate,
): Promise<NotificationPreferences> {
  const response = await api.patch<NotificationPreferences>(
    "/api/v1/notification-preferences",
    data,
  );
  return response.data;
}

File: backend/app/api/routes/notification_preferences.py

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user_id
from app.core.database import get_db

from app.schemas.notification_preference import (
    NotificationPreferenceResponse,
    NotificationPreferenceUpdate,
)

from app.services.notification_preference import (
    get_or_create_preferences,
    update_preferences,
)


router = APIRouter(
    prefix="/api/v1/notification-preferences",
    tags=["Notification Preferences"],
)


# ============================================================
# GET NOTIFICATION PREFERENCES
# ============================================================

@router.get(
    "",
    response_model=NotificationPreferenceResponse,
)
def get_notification_preferences(
    db: Session = Depends(get_db),

    current_user_id: str = Depends(
        get_current_user_id
    ),
):
    preferences = get_or_create_preferences(
        db,
        current_user_id,
    )

    return NotificationPreferenceResponse(
        task_assigned=preferences.task_assigned,
        status_changed=preferences.status_changed,
        task_created=preferences.task_created,
        task_updated=preferences.task_updated,
        mention=preferences.mention,
        comment_added=preferences.comment_added,
    )


# ============================================================
# UPDATE NOTIFICATION PREFERENCES
# ============================================================

@router.patch(
    "",
    response_model=NotificationPreferenceResponse,
)
def update_notification_preferences(
    data: NotificationPreferenceUpdate,

    db: Session = Depends(get_db),

    current_user_id: str = Depends(
        get_current_user_id
    ),
):
    preferences = update_preferences(
        db,
        current_user_id,
        data,
    )

    return NotificationPreferenceResponse(
        task_assigned=preferences.task_assigned,
        status_changed=preferences.status_changed,
        task_created=preferences.task_created,
        task_updated=preferences.task_updated,
        mention=preferences.mention,
        comment_added=preferences.comment_added,
    )